### PR TITLE
GitLab: fetch media library images through API

### LINKED ISSUE #1436: [Bug] CMS on Gitlab not displaying photos in media assets
I just setup Netlify CMS on my site on Gitlab, I noticed that in the admin, the media page doesn't show my pictures.
screen shot 2018-06-14 at 6 56 22 am

When I try to view the image, I noticed the url isn't encoded properly:
screen shot 2018-06-14 at 6 56 35 am

Steps to reproduce:
- Add Netlify CMS to existing Gitlab site (Not using NPM)
- View the Media assets

**CMS Version**
Netlify CMS version: 1.9.0
Browser version: Chrome 67

**Here's my config file**
```yaml
backend:
  name: gitlab
  repo: namzo/ernestojeh.com
  branch: master
  site_domain: ernestojeh.com

display_url: https://ernestojeh.com
media_folder: "img/post" # Media files will be stored in the repo under images/uploads

collections:
  - name: "writing" # Used in routes, e.g., /admin/collections/blog
    label: "Writing" # Used in the UI
    folder: "_posts" # The path to the folder where the documents are stored
    create: true # Allow users to create new documents in this collection
    slug: "{{year}}-{{month}}-{{day}}-{{slug}}" # Filename template, e.g., YYYY-MM-DD-title.md
    fields: # The fields for each document, usually in front matter
      - {label: "Layout", name: "layout", widget: "hidden", default: "post"}
      - {label: "Title", name: "title", widget: "string"}
      - {label: "Publish Date", name: "date", widget: "datetime"}
      - {label: "Body", name: "body", widget: "markdown"}
```
Thanks.

### LINKED ISSUE #1546: Images from Media Assets are not fetched correctly giving 403 when using bitbucket
**- Do you want to request a *feature* or report a *bug*?**
*bug*

**- What is the current behavior?**
image
In my local sandbox attempt to fetch images give 403 for each image request, while other stuff (non images) are fetched correctly. Image urls have wrong url containing guid instead of "master". 
api.bitbucket.org/2.0/repositories/bomzj/house-fresh/src/**84cf67837016f94a5f00e809d62ca2faf7dabca7**/static/img/houseFreshLogo128x128.svg
VS
All other fetch requests have "master" instead of guid e.g. api.bitbucket.org/2.0/repositories/bomzj/house-fresh/src/**master**/static/css/main.css 

In postman, for testing purposes,  I replaced that guid with "master" and got the image.

**- If the current behavior is a bug, please provide the steps to reproduce.**
1) use bitbucket as backend in config.yml
2) log in to a site 
3) open Media assets popup

**- What is the expected behavior?**
Images should display without any issues.

**- Please mention your versions where applicable.**
Netlify CMS version: 2.0.6
Browser version: Chrome 68
Node.JS version: 8.9
Operating System: Windows 10

**- Please link or paste your `config.yml` below if applicable.**
backend:
  name: bitbucket
  repo: private info

media_folder: "static/img"

### LINKED ISSUE #1608: GitLab with Git Gateway: media loading and uploading broken
**- Do you want to request a *feature* or report a *bug*?**
bug

**- What is the current behavior?**
Quoting @DavidRayner from https://github.com/netlify/netlify-cms/issues/1565#issuecomment-412331348:

> I still have this issue with my site which was based off the hugo starter
> I will try to reproduce with the hugo starter upgraded to v2.x
>
> image

**- If the current behavior is a bug, please provide the steps to reproduce.**
Quoting @DavidRayner from https://github.com/netlify/netlify-cms/issues/1565#issuecomment-412333598:

> I believe this issue still exists on netlify-cms 2.0.9
>
> I made a Netlify site using GitHub and https://github.com/DavidRayner/one-click-hugo-cms. I was able to upload media without any problem.
> 
> I made another Netlify site using GitLab and https://gitlab.com/davidrayner/one-click-hugo-cms (an import of the working GH repo) - and I was not able to upload media.

**- What is the expected behavior?**
Media loads and can be uploaded successfully.

**- Please mention your versions where applicable.**

Netlify CMS version: 2.0.9
Browser version:

Node.JS version:
Operating System:

**- Please link or paste your `config.yml` below if applicable.**

### PR DESCRIPTION
**Pull request by Benaiah**
Fetches media library images through the GitLab API, allowing images to be viewed in the media library when using a GitLab repo that is private or using git-gateway. NOTE: this does not solve image display in the _editor_ for those cases, just in the media library.

Closes #1546.
Closes #1436.
Closes #1608.

### CODE REVIEW

**Comment by verythorough** [comment]
Deploy preview for *netlify-cms-www* ready!

Built with commit 988e069e2bfc88891e774aed9e79b2f8cf88f07d

https://deploy-preview-1433--netlify-cms-www.netlify.com

**Comment by verythorough** [comment]
Deploy preview for *cms-demo* ready!

Built with commit 988e069e2bfc88891e774aed9e79b2f8cf88f07d

https://deploy-preview-1433--cms-demo.netlify.com

File: src/components/MediaLibrary/MediaLibrary.js [hunk 1]
```diff
@@ -316,7 +374,12 @@ class MediaLibrary extends React.Component {
                   <div className="nc-mediaLibrary-cardImage-container">
                     {
                       file.isViewableImage
-                        ? <img src={file.url} className="nc-mediaLibrary-cardImage"/>
+                        ? <CardImage
+                          isVisible={true}
```
Comment (erquhart) [inline src/components/MediaLibrary/MediaLibrary.js]: Doesn't look like we're using this.

File: src/components/MediaLibrary/MediaLibrary.js [hunk 2]
```diff
@@ -24,6 +25,60 @@ import { Icon } from 'UI';
...
+        }
+      });
+    }
+  }
+
+  render() {
+    const { imageURL, isFetching } = this.state;
+
+    if (imageURL === "" && !isFetching) {
+      this.loadImage();
```
Comment (erquhart) [inline src/components/MediaLibrary/MediaLibrary.js]: We're calling `setState` during the render - we need to move this state management to Redux.

File: src/backends/gitlab/API.js [hunk 3]
```diff
@@ -32,12 +32,36 @@ export default class API {
...
+  responseFormats = fromJS({
+    json: async res => {
+      const contentType = res.headers.get("Content-Type");
+      if (contentType !== "application/json" && contentType !== "text/json") {
+        throw new Error(`${ contentType } is not a valid JSON Content-Type`);
+      }
+      return res.json();
+    },
+    text: async res => res.text(),
+    blob: async res => res.blob(),
```
[the remark this answers is not in the document]
Comment (Benaiah) [reply src/backends/gitlab/API.js]: I put the `async` there to make it more obvious that these return Promises, but in retrospect this seems to introduce more confusion than it solves.

File: src/components/MediaLibrary/MediaLibrary.js (hunk 1 above)
Comment (Benaiah) [reply src/components/MediaLibrary/MediaLibrary.js]: Good catch - this is leftover from when I was only displaying visible images in the media library (c.f. https://github.com/benaiah/netlify-cms/tree/experimental-media-library-scrolling).

File: src/backends/gitlab/API.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (erquhart) [reply src/backends/gitlab/API.js]: Makes sense - inline comment would work too.

File: src/components/MediaLibrary/MediaLibrary.js (hunk 1 above)
Comment (erquhart) [reply src/components/MediaLibrary/MediaLibrary.js]: Yeah figured as much.

**Comment by tech4him1** [comment]
@Benaiah We should technically be revoking the object URL when it is no longer in use (i.e. the file was deleted). Is that easy enough to add?

**Comment by tech4him1** [comment]
This will fix media viewing *and* uploading for GitLab sites using Git Gateway for authentication (#1608).

**Comment by kokikillara** [comment]
please fix this bug

**Comment by Benaiah** [comment]
@kokikillara thanks for the review!

@tech4him1 I've updated the PR so that display URLs will be revoked on deleting a file - they aren't cached for longer than a single session anyway.

@erquhart just as a ping, you still have a blocking review on this PR.

**Comment by Benaiah** [comment]
Rebased onto the current master, please review ASAP so we can merge this and get bitbucket and gitlab working.

**Comment by tech4him1** [comment]
Awesome, thanks for plugging through this! There are a couple lint errors right now:
```
/packages/netlify-cms-core/src/actions/mediaLibrary.js
  176:27  error  'getState' is defined but never used       no-unused-vars
  182:15  error  'blob' is assigned a value but never used  no-unused-vars
/packages/netlify-cms-core/src/components/MediaLibrary/MediaLibraryCardGrid.js
  40:30  error  'idx' is defined but never used  no-unused-vars
/packages/netlify-cms-core/src/components/MediaLibrary/MediaLibraryCardImage.js
  2:8  error  'PropTypes' is defined but never used  no-unused-vars
```

**Comment by Benaiah** [comment]
@tech4him1 thanks for the catch - looks like I need to fix my editor linking plugin.

**Comment by Benaiah** [comment]
Lint errors should be fixed now.

**Pull request merged**